airflow/models/serialized_dag.py

```diff
@@ -87,7 +101,10 @@ class SerializedDagModel(Base):
     dag_hash = Column(String(32), nullable=False)
```

Comment:
I see in #42690 that you're planning to propagate this everywhere. Since this is a straight up hash of the serdag, I'm starting to wonder if we'd be better off with a separate "dag_version" entity, so that:


```
DagModel  -> DagVersion -> [SerializedDagModel, DagCode]
```

Today they are the same, but fundamentally they aren't.

Reply:
> I see in #42690 that you're planning to propagate this everywhere. Since this is a straight up hash of the serdag, I'm starting to wonder if we'd be better off with a separate "dag_version" entity, so that:
> 
> ```
> DagModel  -> DagVersion -> [SerializedDagModel, DagCode]
> ```
> 
> Today they are the same, but fundamentally they aren't.

The dag_hash is already in use everywhere, and it looks like there are no more places to add it except the TIHistory. There's a `dagrun.dag_hash` which the scheduler uses to create dagrun and also [verify integrity](https://github.com/apache/airflow/blob/db6eb16631d398ab6287b7991da1d571aa591d15/airflow/jobs/scheduler_job_runner.py#L1709) of the DAG and in the #42690, we just needed to record the hash that the task instance ran with. I think that would be all we will do in the PR aside from tests. 

We already have all we need to determine the dag version a task instance ran with, and it looks like a separate DagVersion table will have more queries with joins, etc., as I understand it. WDYT?

Reply:
Another proposal: Instead of a separate versioning table, track the serialized dag version used by dagrun and taskinstance by having a foreignkey between the models and serialized dag model. Currently, this is tracked with only dag_hash but now, dag_hash can be duplicated due to the versioning. This could also allow the scheduler to run with a specific version fed through the UI. Scheduler is currently running with the latest serdag, when it changes, the scheduler change to the updated serdag. We can make it stick to one version until the end of the run.
I will update #42690 with this proposal so you can see what I mean.